entoas/extension.go

```diff
@@ -0,0 +1,196 @@
+package entoas
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"os"
+	"path/filepath"
+
+	"entgo.io/contrib/entoas/policy"
+	"entgo.io/contrib/entoas/spec"
+	"entgo.io/ent/entc"
+	"entgo.io/ent/entc/gen"
+)
+
+type (
+	// Config provides global metadata for the generator. It is injected into the gen.Graph.
+	Config struct {
+		// DefaultPolicy defines the default policy for endpoint generation.
+		// It is used if no policy is set on a (sub-)resource.
+		// Defaults to PolicyExpose.
+		DefaultPolicy policy.Policy
+		// Whether or whether not to generate simple models instead of one model per endpoint.
+		SimpleModels bool
+	}
+	// Extension implements entc.Extension interface for providing OpenAPI Specification generation.
+	Extension struct {
+		entc.DefaultExtension
+		config    *Config
+		mutations []MutateFunc
+		out       io.Writer
+	}
+	// ExtensionOption allows managing Extension configuration using functional arguments.
+	ExtensionOption func(*Extension) error
+	// MutateFunc defines a closure to be called on a generated spec.
+	MutateFunc func(*gen.Graph, *spec.Spec) error
+)
+
+// NewExtension returns a new entoas extension with default values.
+func NewExtension(opts ...ExtensionOption) (*Extension, error) {
+	ex := &Extension{config: &Config{DefaultPolicy: policy.Expose}}
+	for _, opt := range opts {
+		if err := opt(ex); err != nil {
+			return nil, err
+		}
+	}
+	return ex, nil
+}
+
+// Hooks of the Extension.
+func (ex *Extension) Hooks() []gen.Hook {
+	return []gen.Hook{ex.generate}
+}
+
+// Annotations of the extensions.
+func (ex *Extension) Annotations() []entc.Annotation {
+	return []entc.Annotation{ex.config}
+}
+
+// DefaultPolicy sets the default ExclusionPolicy to use of none is given on a (sub-)schema.
+func DefaultPolicy(p policy.Policy) ExtensionOption {
+	return func(ex *Extension) error {
+		ex.config.DefaultPolicy = p
+		return nil
+	}
+}
+
+// Mutations adds the given mutations to the spec generator.
+//
+// A MutateFunc is a simple closure that can be used to edit the generated spec. It can be used to add custom endpoints
+// or alter the spec in any other way.
```

Comment:
```suggestion
// Mutations adds the given mutations to the spec generator.
//
// A MutateFunc is a simple closure that can be used to edit the generated spec.
// It can be used to add custom endpoints or alter the spec in any other way.
```